### ISSUE TITLE
Refactoring a base module

### ISSUE THREAD

**Issue by leamas** [Rasbats/shipdriver_pi#144]
Hi,

I have been thinking about the future work with the stuff now enshrined in the shipdriver plugin. 

The good part seems to be that we now have two directories with stable stuff: cmake/ and ci/. These seems possible to use as-is in more or less all plugins we have encountered so far. Or?

We also have a set of templates like CMakelists.txt, .circleci/config, appveyor.yml etc which can be adapted to new plugins without too much work. We *could* make them more generic, but the price would be more complexity. The idea is that it's better to keep these simple.

The bad part is how the files in cmake/ and ci/ are maintained. Any update of these files causes manual, error-prone and boring work when they should be replicated in all plugins using them. This issue is about this.

I have the idea that we could move cmake/ and ci/ to a separate project -- let's call it base-plugin. And that this new project is incorporated into all plugins using it as a git submodule. This would give the  following layout

```
  shipdriver
    |
    | --- CMakeLists.txt
    |
    | --- plugin-base (submodule)
            |
            | ---   cmake/
            |
            | ---   ci/


```

Yes, some paths will change so there is a delta. But I don't think it's that complicated. OTOH, with this layout changes in the plugin-base project can be imported with a simple `git submodule update --remote plugin-base`. Which is simple enough.

Note that possible changes in plugin-base does not destabilize any plugin. No changes are pulled into the plugin until we update the submodule in the plugin like above.

Bottom line: we would make maintenance much simpler to the price of some added complexity. 

Thoughts?

**Comment by Rasbats** [comment]
> Thoughts?

My focus has always been:

1.  to produce a plugin that "just works".
2.  to make it as easy as possible for a newcomer to add their ideas, with a new plugin.

I have used ShipDriver as the base for a couple of legacy plugins and the 'conversion' process is easy, using the changes that you have made. Witness my cheat sheet.

At the moment we can take a legacy plugin, use that code to test new ideas and then move the plugin to a managed version. Can a workflow be developed so that developers like me, who have a limited understanding of Git or CMake, can use the workflow?

This may not be the right place for suggesting it but ATM we have your workflow and Jon Gough's. As a first step can these be brought together? The developer manual is a total mess for someone looking to create a plugin.

If ./cmake and ./ci are stable how does moving them to a submodule help to create a new plugin? Is it not better to have a plugin like Jon's testplugin for testing changes? Or continue to use ShipDriver as a testbed? Or I can put together a simple demo plugin that uses one or two of the API functions from OpenCPN?

> simple git submodule update --remote plugin-base. Which is simple enough.

For me a copy-paste of two folders, cmake and ci is simpler :)

**Comment by leamas** [comment]
> For me a copy-paste of two folders, cmake and ci is simpler :)

Well, in that case we should drop this idea. I have sort come to the same conclusion after sleeping a night. on it.

>  Witness my cheat sheet.

"blushes" -- I lost the link. Could you please...

>  Or continue to use ShipDriver as a testbed

Yes, let's do that.

> As a first step can these be brought together?

From a technical POV, yes. Unfortunately, there are personal aspects here which makes it hard for me to push such a thing. It *might* work if you do it, but to be honest I doubt it.

> The developer manual is a total mess for someone looking to create a plugin.

Havn't looked at it, but  it would certainly not be a surprise. Let's develop the documentation in your wiki so it works for new devs. When we think it is in shape, let Rick just put a pointer to it as an alternative workflow.

**Comment by Rasbats** [comment]
Cheat Sheet (included in the managed_plugins wiki):
[https://github.com/Rasbats/managed_plugins/blob/main/cheatsheet/Managed.Workflow.Cheat.Sheet.pdf](https://github.com/Rasbats/managed_plugins/blob/main/cheatsheet/Managed.Workflow.Cheat.Sheet.pdf)

Developer Manual:
[https://opencpn.org/wiki/dokuwiki/doku.php?id=opencpn:developer_manual:pi_installler_build_deploy](https://opencpn.org/wiki/dokuwiki/doku.php?id=opencpn:developer_manual:pi_installler_build_deploy)

**Comment by leamas** [comment]
Thanks, sorry for short memory. I filed a request to be collaborator in the wiki: https://github.com/Rasbats/managed_plugins/issues/1

**Comment by leamas** [comment]
Anyway, let's drop this idea. Closing

**Comment by leamas** [comment]
off-topic: You did find this, right: https://circleci.com/docs/2.0/oss/

**Comment by Rasbats** [comment]
Sorry... missed that one.  'Build forked pull requests' now set to 'On'. Could you make a small PR to test?
